File: src/info/meodinger/LabelPlusFX/Util/CTree.java
package info.meodinger.LabelPlusFX.Util;

import javafx.scene.control.TreeItem;

/**
 * @author Meodinger
 * Date: 2021/5/29
 * Location: info.meodinger.LabelPlusFX.Util
 */
public class CTree {

    public static <T> TreeItem<T> getRootOf(TreeItem<T> item) {
        if (item == null) return null;

        TreeItem<T> root = item;
        while (root.getParent() != null) root = root.getParent();
        return root;
    }

    public static void expandAll(TreeItem<?> item) {
        if (item == null) return;

        if (item.getChildren().size() > 0) {
            item.setExpanded(true);
            for (TreeItem<?> i : item.getChildren()) {
                expandAll(i);
            }
        }
    }
}

File: src/main/kotlin/info/meodinger/lpfx/util/CTree.kt
package info.meodinger.lpfx.util.tree

import javafx.scene.control.TreeItem

/**
 * Author: Meodinger
 * Date: 2021/7/29
 * Location: info.meodinger.lpfx.util
 */
val <T> TreeItem<T>.root: TreeItem<T>
    get() {
        var root: TreeItem<T> = this
        while (root.parent != null) root = root.parent
        return root
    }

fun TreeItem<*>.expandAll() {
    if (this.children.size > 0) {
        this.isExpanded = true
        for (i in this.children) {
            i.expandAll()
        }
    }
}